Images are now registered via a RANSAC process at the start of Regist…

ste7e opened the pull request:
…erStarsAndFit (if the AutoFocusOptions.UseRANSAC option is true)

ste7e committed:
Rebasing changes to add RANSAC Alignment and non-trees-based matching

ste7e committed:
Rebasing changes to add RANSAC Alignment and non-trees-based matching

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Utility/RANSACRegistration.cs:
```diff
@@ -0,0 +1,196 @@
+                double minDistance = double.MaxValue;
+
+                foreach (var targetStar in targetStars) {
+                    double distance = Math.Sqrt(
```
I suggest using the Quadtree implementation I have already for this. It'll make this filtering much faster

ghilios replied:
It's the use of KdTree

ste7e replied:
No good reason to do that sqrt in there just for comparing distances.  That's a mistake on my part.  Glad you selected that snippet otherwise I wouldn't have spotted that

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Utility/RANSACRegistration.cs:
```diff
@@ -0,0 +1,196 @@
+                        Math.Pow(refStar.Y - targetStar.Y, 2));
+
+                    // Filter by magnitude difference
+                    if (Math.Abs(refStar.Magnitude - targetStar.Magnitude) <= magnitudeDiffThreshold &&
```
I'd expect magnitude to vary by frame - dimmer as the star gets more diffuse. What inspired you to add this, and how are you going about picking the threshold?

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Utility/RANSACRegistration.cs:
```diff
@@ -0,0 +1,196 @@
+            return bestTransform;
+        }
+
+        private static SimilarityTransform FitSimilarityTransform(List<Point2D> srcPoints, List<Point2D> dstPoints) {
```
Also try out estimateAffinePartial2D in OpenCV. HocusFocus already imports it with a .NET wrapper, so trying it out should be fairly straightforward. One thing I liked to do while implementing this was have multiple approaches behind feature flags in the options, and then just run it multiple times with different settings to see how performance actually is

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Utility/RANSACRegistration.cs:
```diff
@@ -0,0 +1,196 @@
+            }
+
+            // Solve least-squares problem
+            var x = A.Solve(b);
```
this is pretty cool - nice!

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Utility/RANSACRegistration.cs:
```diff
@@ -0,0 +1,196 @@
+
+    internal class RANSACRegistration {
+
+        // Generate putative matches using nearest neighbor
```
this is probably the most reasonable way to go about starting this, since we don't expect rotations. however I do expect scale differences depending on optics. for example, SCTs actually have focal length change as focus shifts, so stars get bigger and FOV shrinks. This use case is the main motivator for doing this.

We probably want to have some way to toggle between RANSAC and basic global registration using nearest neighbors (the current approach) so we can choose empirically between them

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Controls/ILNSceneControlBase.xaml.cs:
```diff
@@ -164,6 +164,7 @@ protected virtual void OnClearScene() {
         }
 
         protected void UpdateScene() {
+            return;
```
what happened here? is this because of not having an ILNumerics license? You should at least be able to use a locally built version, if I recall correctly.

ghilios replied:
Something I'd like to do in the future is replace this library with something that is free and unlicensed. I couldn't get alternates to work when I first built this. I'm very annoyed with his licensing model

ste7e replied:
Yes, it was just easier to skip the ILNumerics stuff for testing

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Inspection/SensorModel.cs:
```diff
@@ -336,78 +332,200 @@ private RegistrationAndFitResult RegisterStarsAndFit(
-                }
+                RegisteredStar[] registeredStars;
+
+                if (autoFocusOptions.UseRANSAC) {
```
probably worth refactoring the registration approach into different functions

ste7e replied:
Agreed.  I wasn't thinking it would stay like this, it was just a quick and dirty way of developing the RANSAC stuff while leaving the other method in place for comparison

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Inspection/SensorModel.cs:
```diff
@@ -336,78 +332,200 @@ private RegistrationAndFitResult RegisterStarsAndFit(
+                        .OrderBy(s => s.X + (s.Y * imageSize.Width));
+                    double brightnessMinFactor = 0;
+                    double maxBrightnessDiff = 0.5d;
+                    double brightnessMinimumRef = allDetectedStars[minHfrIndex].StarDetectionResult.StarList.Average(s => s.MaxBrightness) * brightnessMinFactor;
```
this will always be 0?

ste7e replied:
The brightnessMinimumref was something I was trying but it didn't really make much difference with the data I was testing against so zeroed if for now

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Inspection/SensorModel.cs:
```diff
@@ -336,78 +332,200 @@ private RegistrationAndFitResult RegisterStarsAndFit(
+                        var (putativeSrc, putativeDst) = RANSACRegistration.GeneratePutativeMatches(
+                            theseStars.Where(s => s.Magnitude > brightnessMinimum).ToList(),
+                            targetStars.Where(s => s.Magnitude > brightnessMinimumRef).ToList(),
+                            maxDistance: 50, magnitudeDiffThreshold: maxBrightnessDiff);
```
look in InspectorVM where I generate registered, annotated images. This was really helpful for debugging. You could extend it to show other information around each star, so you can see visually what stars were matched with eachother, and later also which ones were excluded due to bad fits (or outlier detection when we add that)

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Inspection/SensorModel.cs:
```diff
@@ -336,78 +332,200 @@ private RegistrationAndFitResult RegisterStarsAndFit(
+                        if (i == minHfrIndex) {
+                            continue;
+                        }
+                        foreach (var star in allDetectedStars[i].StarDetectionResult.StarList) {
```
this will be really slow for large star fields. I suggest using the KdTree.

ste7e replied:
OK

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Inspection/SensorModel.cs:
```diff
@@ -336,78 +332,200 @@ private RegistrationAndFitResult RegisterStarsAndFit(
-                            continue;
+                    float minBrightness = 0.05f;
+                    float minMatchProportion = 0.75f;   // proportion of images that must have star for star to be included
+                    int starCount = starDict.Keys.Count(pt => starDict[pt].Count >= allDetectedStars.Count * minMatchProportion);   // only use stars that are matched by 75% of the images
```
this becomes a tall order with SCTs, where large stars become donuts and are no longer detected. Perhaps we can make this a setting for a while to test out different options? Proportion, min number, outlier detection, etc

ste7e replied:
Yep, we need to try it a lot of different cases.  Might well be a case of reiterating with a sliding value until we get a usable result set

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Inspection/SensorModel.cs:
```diff
@@ -336,78 +332,200 @@ private RegistrationAndFitResult RegisterStarsAndFit(
+                    }).ToArray();
+                    stopwatch.RecordEntry("build trees");
+
+                    float searchRadius = autoFocusOptions.UseRANSAC ? 3 : 30;
```
in this else, we already know UseRANSAC is false, right?

ste7e replied:
Yep.  That's a legacy of when I used RANSAC and transformed the images and then used the existing matching routines but with a tighter searchRadius

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/AutoFocus/InspectorVM.cs:
```diff
@@ -782,11 +782,12 @@ private async Task<bool> AnalyzeSavedAutoFocusRun() {
-                            Logger.Error($"Multiple saved images in final directory {finalDirectory}. Continuing without doing exposure analysis");
-                            Notification.ShowError($"Multiple saved images in final directory {finalDirectory}. Continuing without doing exposure analysis");
-*/                        } else {
+                            /*                        } else if (savedFinalAttempt.SavedImages.Count > 1) {
```
should these be uncommented now?

ste7e replied:
Not really.  If we do, that message will appear every time you run AI on a final folder that was created with multiple frames per point.  The way I see it, we have 3 options:
1 we remove those lines in which case the first frame in the final folder will be used
2 we select one of the images to use from the final folder
3 we only take one frame for the final folder even if multiple frames per point is selected

ghilios replied:
ok then let's delete this block entirely instead of comment them out

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/AutoFocus/AutoFocusEngine.cs:
```diff
@@ -926,8 +926,8 @@ private async Task<AutoFocusState> InitializeState(
             CancellationToken token,
-            IProgress<ApplicationStatus> progress) {
-            var autofocusFilter = await SetAutofocusFilter(imagingFilter, token, progress);
+            IProgress<ApplicationStatus> progress, bool forRerun = false) {
```
nit: put parameter on separate line

ste7e replied:
👍

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/AutoFocus/AutoFocusEngine.cs:
```diff
@@ -1618,8 +1618,22 @@ public SavedAutoFocusAttempt LoadSavedFinalAttempt(string path) {
+                            attemptFolder = subFolders.FirstOrDefault();
+                        }
+                    }
+                    if (!attemptMatch.Success)
```
nit: as a matter of style, please always use braces even for single line blocks

ste7e replied:
👍

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/AutoFocus/InspectorOptions.cs:
```diff
@@ -342,5 +348,57 @@ public bool FixedSensorCenter {
+            }
+        }
+
+        private bool useTrees = false;
```
how does performance and results differ for you with/without trees? I'm wondering what our default should be here

ste7e replied:
Yeah, that still needs to be decided... 

On my tests:

Test1 (a bad set of data)
> With Trees R² 0.29 Points in model (PIM) 706 in 43secs (time for Register & Fit)
> Non-trees R² 0.26 PIM 604 in 57secs

Test2 (a multi-frame per point set of data - again bad data (too much sky glow, bad seeing and clouds))
> With Trees R² 1.00 PIM 7 in 14 secs 
> > (I'm surprised about the 1.00 on this one, I was avoiding 1.00 results because it likely means too few points for a good fit - need to recheck the code for that.  The second best iteration was 0.5 with 12 stars)
> Non-trees R² 0.92 PIM 26 in 0.5 secs

Test3 (your large star field dataset)
> With Trees R² 0.96 PIM 2736 in 24secs 
> Non-trees R² 0.98 PIM 5175 in 28secs

Test4 (your SCT dataset)
> With Trees R² 0.63 PIM 18 in 0.34 secs
> Non-trees R² 0.71 PIM 81 in 1.6 secs 

Test5 (a 'normal' dataset with single frames per point)
> With Trees R² 0.88 PIM 72 in 0.5secs 
> Non-trees R² 0.92 PIM 93 in 0.12secs 

So, the non-trees method has slightly better results at the cost of a slight performance hit.

I don't think you'd want both methods in the final release just need to decide which method to use, I guess

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/AutoFocus/InspectorVM.cs:
```diff
@@ -309,9 +310,10 @@ await SensorModel.UpdateModel(
-                if (!String.IsNullOrEmpty(result.SaveFolder)) {
-                    await SaveRegisteredImages(result.SaveFolder, SensorModel.SensorModelResult.RegisteredStars);
-                }
+                if (!forRerun)
```
add braces

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Properties/launchSettings.json:
```diff
@@ -0,0 +1,11 @@
+{
```
please remove this file. shouldn't need this committed

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Utility/RANSACRegistration.cs:
```diff
@@ -0,0 +1,197 @@
+namespace NINA.Joko.Plugins.HocusFocus.Utility {
+
+    public class Point2D {
+        public double X { get; set; }
```
should these setters be private?

ste7e replied:
Yes they should

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Utility/RANSACRegistration.cs:
```diff
@@ -0,0 +1,197 @@
+                foreach (var targetStar in targetStars) {
+                    double x = refStar.X - targetStar.X;
+                    double y = refStar.Y - targetStar.Y;
+                    double distance = x * x + y * y;
```
perhaps call this distance2 so it's clear this is squared?

ste7e replied:
Good call

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Utility/RANSACRegistration.cs:
```diff
@@ -0,0 +1,197 @@
+                    double distance = x * x + y * y;
+
+                    // Filter by magnitude difference
+                    if (distance < bestDistance && distance < maxDistance2 && Math.Abs(refStar.NormalisedBrightness - targetStar.NormalisedBrightness) <= relativeBrightnessDiff
```
minor, but if you move the brightness check to the top of the foreach you could skip some extra computation

ste7e replied:
Another good call

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Utility/RANSACRegistration.cs:
```diff
@@ -0,0 +1,197 @@
+                throw new ArgumentException("At least 2 corresponding points are required.");
+            }
+
+            Random rand = new Random();
```
how about making the RNG a static constant somewhere?

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Utility/RANSACRegistration.cs:
```diff
@@ -0,0 +1,197 @@
+
+            for (int i = 0; i < maxIterations; i++) {
+                // Randomly select 2 points for similarity transform
+                int[] indices = Enumerable.Range(0, srcPoints.Count).OrderBy(x => rand.Next()).Take(2).ToArray();
```
this is a really inefficient way to shuffle, since it sorts the whole thing. Just choose 2 distinct random numbers and get the indexes directly out of the list

ste7e replied:
yeah you're right.  that bit of code just sort of evolved as I was working in the ransac and I never went back to give it another look

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Utility/RANSACRegistration.cs:
```diff
@@ -0,0 +1,197 @@
+                for (int j = 0; j < srcPoints.Count; j++) {
+                    var transformedPoint = transform.Transform(srcPoints[j]);
+                    double distance = Math.Sqrt(
+                        Math.Pow(transformedPoint.X - dstPoints[j].X, 2) +
```
instead of Pow, just multiply twice. Math.Pow uses a Taylor series under the hood IIRC, which is a really inefficient way of doing this really frequent computation

ste7e replied:
Agreed.  I don't need the sqrt either as we're justt finding the nearest and don't actually need the distance

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Utility/RANSACRegistration.cs:
```diff
@@ -0,0 +1,197 @@
+                List<int> inlierIndices = new List<int>();
+                for (int j = 0; j < srcPoints.Count; j++) {
+                    var transformedPoint = transform.Transform(srcPoints[j]);
+                    double distance = Math.Sqrt(
```
can you skip the sqrt?

ste7e replied:
indeed

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/AutoFocus/InspectorOptions.cs:
```diff
@@ -342,5 +348,57 @@ public bool FixedSensorCenter {
             }
         }
+
+        private bool useRANSAC = true;
```
how about we start with ransac off by default, and ask people to use it and give us feedback?

ste7e replied:
Yes agreed

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Inspection/SensorModel.cs:
```diff
@@ -335,131 +349,533 @@ private RegistrationAndFitResult RegisterStarsAndFit(
+                int ransacAligned = 0;
+                if (inspectorOptions.UseRANSAC) {
+                    ransacAligned = AlignStarsWithRANSAC(allDetectedStars, imageSize, stopwatch, minHfrIndex);
+                    if (ransacAligned == 0)
```
use braces

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Inspection/SensorModel.cs:
```diff
@@ -335,131 +349,533 @@ private RegistrationAndFitResult RegisterStarsAndFit(
+                if (inspectorOptions.UseRANSAC) {
+                    ransacAligned = AlignStarsWithRANSAC(allDetectedStars, imageSize, stopwatch, minHfrIndex);
+                    if (ransacAligned == 0)
+                        Trace.WriteLine($"Ransac failed, reverting to non-aligned processing");
```
nit: don't need interpolation if there are no variables

ste7e replied:
oops

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Inspection/SensorModel.cs:
```diff
@@ -335,131 +349,533 @@ private RegistrationAndFitResult RegisterStarsAndFit(
+                for (bool retry = true; retry;) {
+                    retry = false;
+                    int rejectionsOnBrightnessDiff;
+                    if (inspectorOptions.UseTrees)
```
use braces

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Inspection/SensorModel.cs:
```diff
@@ -335,131 +349,533 @@ private RegistrationAndFitResult RegisterStarsAndFit(
+                    try {
+                        pfit = FitParaboloidModel(focuserSizeMicrons, finalFocusPosition, imageSize, pixelSize, reg, ct);
+                    } catch (Exception ex) {
+                        Trace.WriteLine($"{ex}");
```
if we fail to fit here, why is retrying the way to go automatically? what changes for the next iteration? how do we ensure it doesn't just keep retrying forever?

ste7e replied:
That was a quick and dirty - I suddenly started getting 'empty sequence' exceptions being thrown - basically too few stars.  When the loop retries it changes the brightness tolerance so the sequence won't be the same on subsequent loops.  I should check that the exception is actually that one - string match on the ex.Message ... urgh, yukk!

What I'll do is create a custom exception for that case instead of a string match.  

To answer the infinite loop question: it'll change the brightness tolerance in the retry and even if we get the same exception for every iteration the loop will end, as normal, when the brightness tolerance results in no rejections

ste7e replied:
I've found why this exception was being thrown now.  It is when the InputEnabledCount is zero the GoodnessOfFit calculation fails.  I've changed GoodnessOfFit now so that it returns a fit of zero if InputEnabledCount is zero.  So, I can get rid of that catch now.  Any exceptions are really exceptions now

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Inspection/SensorModel.cs:
```diff
@@ -335,131 +349,533 @@ private RegistrationAndFitResult RegisterStarsAndFit(
+                        retry = true;
+                    }
+                    if ((pfit != null) && (rejectionsOnBrightnessDiff > 0) && ((pfit.StarsInModel < 10) || (pfit.GoodnessOfFit < 0.6) || (pfit.GoodnessOfFit == 1))) {
+                        Trace.WriteLine($"Only have {pfit.StarsInModel} stars and R2 of {pfit.GoodnessOfFit:#.##} with a MNBD of {maxNormalisedBrightnessDiff:#.##}");
```
this might actually warrant more than a trace statement. Since this is going to retry several times perhaps, we should be able to find this information in the logs for users.

When this is done, what do you think about showing a notification about how many attempts it took?

ste7e replied:
Logging: yes definitely

Information for users, yes we could do that.  It's useful to see that soft of info from StarAlignment PixInsight ... probably a good idea.  How would you want to show that?

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Inspection/SensorModel.cs:
```diff
@@ -335,131 +349,533 @@ private RegistrationAndFitResult RegisterStarsAndFit(
-                            var distance = MathUtility.DotProduct(pointDiff, pointDiff);
-                            queue.Enqueue(new MatchingPair() { SourceIndex = sourceIndex, GlobalIndex = globalNeighborIndex }, distance);
-                        }
+                    if ((bestPfit == null) || ((pfit != null) && ((pfit.GoodnessOfFit > bestPfit.GoodnessOfFit) || (bestPfit.GoodnessOfFit == 1)) && (pfit.GoodnessOfFit < 1))) {
```
this is really hard to read. can you break out some of the conditions to variables?

ste7e replied:
I should be using IsBetterFit in there, actually.  Something I forgot to back and change when I decided it would be easier to read with a helper function

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Inspection/SensorModel.cs:
```diff
@@ -335,131 +349,533 @@ private RegistrationAndFitResult RegisterStarsAndFit(
+                    if (retry) {    // keep going in the same direction
+                        Trace.WriteLine($"this MNBD={maxNormalisedBrightnessDiff:#.##}, direction: {direction}");
+                        if (direction == IterationDirection.Up) {
+                            if (maxNormalisedBrightnessDiff >= 3)
```
add braces

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Inspection/SensorModel.cs:
```diff
@@ -335,131 +349,533 @@ private RegistrationAndFitResult RegisterStarsAndFit(
+                                direction = IterationDirection.Down;
+                            }
                         }
+                        if (retry)
```
add braces

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Inspection/SensorModel.cs:
```diff
@@ -335,131 +349,533 @@ private RegistrationAndFitResult RegisterStarsAndFit(
+        }
+
+        private static bool IsBetterFit(SensorParaboloidModel thisFit, SensorParaboloidModel otherFit) {
+            if (thisFit == null)
```
add braces

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Inspection/SensorModel.cs:
```diff
@@ -335,131 +349,533 @@ private RegistrationAndFitResult RegisterStarsAndFit(
+                    var points = registeredStar.MatchedStars.Select(s => new ScatterErrorPoint(s.FocuserPosition, s.Star.HFR, 0.0d, 0.0d)).ToList();
+                    AlglibHyperbolicFitting fitting;
+                    if (autoFocusOptions.UnevenHyperbolicFitEnabled) {
+                        fitting = HyperbolicUnevenFittingAlglib.Create(this.alglibAPI, points, stepSize, autoFocusOptions.WeightedHyperbolicFitEnabled);
```
this is where we should add outlier detection, right? I could add it to your branch if you like. I think we just need to refactor it so they can share an implementation

ste7e replied:
Yes please

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Inspection/SensorModel.cs:
```diff
@@ -335,131 +349,533 @@ private RegistrationAndFitResult RegisterStarsAndFit(
+                        continue;
+                    }
+
+                    if (fitting.RSquared < 0.90) {
```
should this R-squared fitting threshold be configurable?

ste7e replied:
It propably should.  I haven't actually looked too much as this section.  That's directly from the current release, I think

ghilios replied:
ok I will add it sometime later

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Inspection/SensorModel.cs:
```diff
@@ -335,131 +349,533 @@ private RegistrationAndFitResult RegisterStarsAndFit(
+                    var dataPoint = new SensorParaboloidDataPoint(dataPointX, dataPointY, focuserMicrons, fitting.RSquared);
+                    sensorModelDataPoints.Add(dataPoint);
+                } catch (Exception e) {
+                    Logger.Error(e, $"Failed to calculate hyperbolic at ({registeredStar.RegistrationX}, {registeredStar.RegistrationY}). Error={e.Message}");
```
if this fails, do we consider it a discarded star?

ste7e replied:
Yeah, looks like it.  Again this is old code and not something I've looked at but that's the way it looks to me

ghilios replied:
ok I will make some updates afterwards then

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Inspection/SensorModel.cs:
```diff
@@ -335,131 +349,533 @@ private RegistrationAndFitResult RegisterStarsAndFit(
+            int imagesAligned = 0;
+
+            var targetStars = allDetectedStars[minHfrIndex].StarDetectionResult.StarList
+                .Select(s => new Point2D(s.Position.X, s.Position.Y/*, ((HocusFocusDetectedStar)s).NormalisedBrightness*/))
```
should the brightness be added back?

ste7e replied:
No that comment (and the code within it) can go.  

I was experimenting with adding the brightness as a third dimension but it wasn't consistent enough across different images.  I needed to add a multiplier to try to balance the difference in brightness to the difference in distance so that two similarly bright stars at the same position on one axis but completely different position on the other axis didn't look like a better match than two stars at exactly the same position with different brightnesses.

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Inspection/SensorModel.cs:
```diff
@@ -335,131 +349,533 @@ private RegistrationAndFitResult RegisterStarsAndFit(
+                    }
+                    imagesAligned++;
+                } catch (Exception ex) {
+                    Trace.WriteLine($"Image {i}: Error: {ex.Message}");
```
if this happens - what next? shouldn't this whole procedure fail?

ste7e replied:
Not really, we can one or two images fail alignment without abandonning the whole dataset.  

If any the images fail to align the RegisterStarsAndFit methof will resort to using the same radius for matches as if the ransac alignment was switched off.  

I've just noticed that I log that that is happening to Trace but only if ALL images failed alignment.  It should be if ANY do.  Also we probably want to send that to the logger instead of Trace

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Inspection/SensorModel.cs:
```diff
@@ -335,131 +349,533 @@ private RegistrationAndFitResult RegisterStarsAndFit(
+                .OrderBy(s => s.X + (s.Y * imageSize.Width));
+            double maxNormalisedBrightnessDiff = .1d;
+            int maxDistanceForPutativeMatch = 25;
+            for (int i = 0; i < allDetectedStars.Count; ++i) {
```
it took me a while to understand this. so this top loop is iterating through each frame? might be worth renaming this variable

ste7e replied:
OK, renamed to imageIndex.

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Inspection/SensorModel.cs:
```diff
@@ -335,131 +349,533 @@ private RegistrationAndFitResult RegisterStarsAndFit(
+                if (i == minHfrIndex) {
+                    continue;
+                }
+                var theseStars = allDetectedStars[i].StarDetectionResult.StarList
```
why is ordering them important here? this looks to be sorted top down first, then left to right?

ste7e replied:
Early on I needed to sort the stars to get good putative matches.  The way I do the putative matches now doesn't need the stars in order so you're right, the expensive sort can go :)

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Inspection/SensorModel.cs:
```diff
@@ -335,131 +349,533 @@ private RegistrationAndFitResult RegisterStarsAndFit(
+
+            stopwatch.RecordEntry("NonTree-based registration");
+
+            if (minBrightness > 0)
```
add braces

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Inspection/SensorModel.cs:
```diff
@@ -335,131 +349,533 @@ private RegistrationAndFitResult RegisterStarsAndFit(
-                stopwatch.RecordEntry("fitcurves");
-                if (discardedStarCount > 0) {
-                    Logger.Warning($"Discarded {discardedStarCount} stars during sensor modeling due to poor fits");
+        private (RegisteredStar[], int) MatchStarsWithoutKdTree(
```
do these two approaches produce the same results? My expectation is that the registration will return stars that are closest to one another, which should be deterministic. This would make for a good test case.

additionally, does the non-tree approach have an advantage? is it faster or does it provide some better accuracy?

ste7e replied:
NonTrees does give slightly better results - better R² with more data points - but does take a little more time

This is because in the nonTrees version we find stars which are on more images (starts looking for stars on 100% of images and reduces that percentage until we have enough stars) in all regions

ste7e committed:
Remove ILNumberics bypass hacks

ste7e committed:
Remove ILNumberics bypass hacks

ste7e committed:
Added option SaveImagesOnReruns
Fixed bugs:
set ImageIndex in MatchStarsWithoutKdTree
use width and height ratios in CreateFullRegionSet

ste7e committed:
Added option SaveImagesOnReruns
Fixed bugs:
set ImageIndex in MatchStarsWithoutKdTree
use width and height ratios in CreateFullRegionSet

ste7e committed:
Added Robust RANSAC (similar triangles)

ste7e committed:
Added Robust RANSAC (similar triangles)

ste7e committed:
Improved triangle generation and matching but getting inconsistent results -R2 and Points in Model changes on each run

ste7e committed:
Improved triangle generation and matching but getting inconsistent results -R2 and Points in Model changes on each run

ste7e committed:
Squashed commit of the following:

commit 5f18c6804b41e971e9575c074992358da07509ea
Author: ste7e <75219076+ste7e@users.noreply.github.com>
Date:   Mon Nov 3 17:50:21 2025 +0000

    Added MaxStarsPerRegion option
    GeneratePutativeMatchesUsingTriangles now builds cosSimalarity for location and brightness seperately
    Now draws circles to show position of reference star on annotated images

commit fcdde84e0d25efeb4f05def7ecdc4e21b23d2f0f
Author: ste7e <75219076+ste7e@users.noreply.github.com>
Date:   Sun Nov 2 18:39:48 2025 +0000

    Fixed registserdStarsInregion loop

commit df06f69f1dccfe7f91401b524dcc5016d8a3613e
Author: ste7e <75219076+ste7e@users.noreply.github.com>
Date:   Sat Nov 1 17:34:16 2025 +0000

    Fixed stuff to get consistent results but R2 isn't now as good as it was - 11 sec test was .82 now .54 but with loads more points - so we're including bad matches somewhere...

commit d4769b2b2fbc3a05a021cce8c8d961250b4955f8
Author: ste7e <75219076+ste7e@users.noreply.github.com>
Date:   Thu Oct 30 10:32:06 2025 +0000

    Investigating bad ransac results.  Temporarily made it non-random, need to do a proper fix for that

ste7e committed:
Squashed commit of the following:

commit 5f18c6804b41e971e9575c074992358da07509ea
Author: ste7e <75219076+ste7e@users.noreply.github.com>
Date:   Mon Nov 3 17:50:21 2025 +0000

    Added MaxStarsPerRegion option
    GeneratePutativeMatchesUsingTriangles now builds cosSimalarity for location and brightness seperately
    Now draws circles to show position of reference star on annotated images

commit fcdde84e0d25efeb4f05def7ecdc4e21b23d2f0f
Author: ste7e <75219076+ste7e@users.noreply.github.com>
Date:   Sun Nov 2 18:39:48 2025 +0000

    Fixed registserdStarsInregion loop

commit df06f69f1dccfe7f91401b524dcc5016d8a3613e
Author: ste7e <75219076+ste7e@users.noreply.github.com>
Date:   Sat Nov 1 17:34:16 2025 +0000

    Fixed stuff to get consistent results but R2 isn't now as good as it was - 11 sec test was .82 now .54 but with loads more points - so we're including bad matches somewhere...

commit d4769b2b2fbc3a05a021cce8c8d961250b4955f8
Author: ste7e <75219076+ste7e@users.noreply.github.com>
Date:   Thu Oct 30 10:32:06 2025 +0000

    Investigating bad ransac results.  Temporarily made it non-random, need to do a proper fix for that

ste7e committed:
Braces

ste7e committed:
Braces

ste7e committed:
Removed ransac and trees from options page as they're now under options -> experimental on the AI page

ste7e committed:
Removed ransac and trees from options page as they're now under options -> experimental on the AI page

ste7e committed:
merged changes from origin

ste7e committed:
removed launchSettings

ste7e committed:
Switvhed from DitProduct to SumOfSquaresOfDifferences

ste7e committed:
Use star count to select reference image for registration and matching
Output StarAlignment images
Fallback to relaxed cosineSimilarity if too few triangles are matched
Show X marks on Registered images to indicate position of star after RANSAC
Corrected location of ellipse around registered stars

ste7e committed:
Use Affine Transform to cope with different scaling in x and y

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/AutoFocus/AutoFocusEngine.cs:
```diff
@@ -1390,8 +1391,8 @@ private async Task<AutoFocusResult> RerunImpl(AutoFocusEngineOptions options, Sa
-                    var imageNumber = focuserPositionGroup.Max(g => g.ImageNumber);
-                    var files = focuserPositionGroup.Where(g => g.ImageNumber == imageNumber).OrderBy(g => g.FrameNumber).ToList();
+                    //var imageNumber = focuserPositionGroup.Max(g => g.ImageNumber);
+                    var files = focuserPositionGroup./*Where(g => g.ImageNumber == imageNumber).*/OrderBy(g => g.FrameNumber).ToList();
```
remove this now?

ste7e replied:
I think we still need to change - with the comments removed.  The original code just sets `files` to a list of one item - the latest image taken at that focus position.  The change sets it to the list of image at the focus position.

So just: 
`var files=focuserPositionGroup.OrderBy(g => g.FrameNumber).ToList()`

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/AutoFocus/AutoFocusEngine.cs:
```diff
@@ -1587,7 +1588,7 @@ public AutoFocusEngineOptions GetOptions(SavedAutoFocusAttempt savedAttempt = nu
                 AutoFocusTimeout = TimeSpan.FromSeconds(autoFocusOptions.AutoFocusTimeoutSeconds),
                 AutoFocusInitialOffsetSteps = profileService.ActiveProfile.FocuserSettings.AutoFocusInitialOffsetSteps,
-                AutoFocusStepSize = savedAttempt?.StepSize ?? profileService.ActiveProfile.FocuserSettings.AutoFocusStepSize,
+                AutoFocusStepSize = (savedAttempt?.StepSize > 0) ? savedAttempt.StepSize.Value : profileService.ActiveProfile.FocuserSettings.AutoFocusStepSize,
```
just to double check: if StepSize is null, will this bool condition throw?

ste7e replied:
Yeah, it'd throw a null exception.  TBH, I can't remember making this change.  i guess I found it was either negative to indicate a default value or positive but it is a nullable int so you're right to flag this.  I'll change it to handle null.

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/AutoFocus/DataTemplates.xaml:
```diff
@@ -1476,6 +1477,193 @@
+                                    <CheckBox
+                                        IsEnabled="{Binding InspectorOptions.UseRANSAC}"
+                                        Margin="5,0,0,0"
+                                        IsChecked="{Binding InspectorOptions.UseRANSACTriangles}" />
```
any reason to keep the first RANSAC approach anymore? What do you think about just keeping the triangles one now?

ste7e replied:
I think probably not.  My original feeling was that it would be slower with triangles and not always necessary.  But now I think it's probably best to have only the triangles option.  I'll probably leave the old code in there just hidden and unused if that's ok - just in case we want to back to it for some reason.  If you prefer it to be out of the code altogether, that's ok though,  Either way I'll remove the option flag etc

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/AutoFocus/DataTemplates.xaml:
```diff
@@ -1476,6 +1477,193 @@
+                                    Grid.Row="1"
+                                    Grid.Column="1"
+                                    Orientation="Horizontal">
+                                    <CheckBox Margin="5,0,0,0" IsChecked="{Binding InspectorOptions.UseAlternativeMatching}" />
```
similar for this - now that we've addressed the squared distance issue, is there any reason to use the alternative matching? can we converge on just the kd trees?

ste7e replied:
Agreed.  Simple is good :)

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Controls/ILNSceneControlBase.xaml.cs:
```diff
@@ -164,6 +164,7 @@ protected virtual void OnClearScene() {
         }
 
         protected void UpdateScene() {
+            return;
```
i'll remove this after we merge it in

ste7e replied:
oh yeah.  I should've controlled that with a condition compile flag really

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Utility/RANSACRegistration.cs:
```diff
@@ -0,0 +1,832 @@
+        }
+
+        public Point2D Transform(Point2D point) {
+            double cosTheta = Math.Cos(Rotation);
```
nit: can you store cos rot and sin rot as private variables so these aren't recalculated for every point?

ste7e replied:
Good spot.  Will do

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Utility/RANSACRegistration.cs:
```diff
@@ -0,0 +1,832 @@
+                for (int i = 0; i < lengths.Count; i++) {
+                    lengths[i] /= shortestLength;
+                }
+                return lengths;
```
does it matter if the lengths returned are in a different order than clockwise due to the sorting?

ste7e replied:
I was doing that so that it would match triangles of the same shape no matter their rotation.  But the clockwise/anti-clockwise thing is a good question.  I'll be matching not just rotated triangles but mirror images, too.  Which is wrong...  might need to rethink this one

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Utility/RANSACRegistration.cs:
```diff
@@ -0,0 +1,832 @@
+                    normPos.Add(new Point2D(
+                        p.X / imageSize.Width,
+                        p.Y / imageSize.Height,
+                        (p.NormalisedBrightness - minBrightness) / (maxBrightness - minBrightness)));
```
there might be a potential divide by zero here if min and max brightness are ever the same

ste7e replied:
Yeah, unlikely but it needs handling.  Will catch it

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Utility/RANSACRegistration.cs:
```diff
@@ -0,0 +1,832 @@
+                return normPos;
+            }
+
+            private double lineLength(Point2D p1, Point2D p2) {
```
nit: name should be lineLengthSquared

ste7e replied:
Yup

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Utility/RANSACRegistration.cs:
```diff
@@ -0,0 +1,832 @@
+            public List<double> NormalizedLengths { get => normalizedLengths; }
+            public List<double> NormalizedBrightnesses { get => normalizedBrightnesses; }
+
+            //public bool IsMatchOnLength(StarTriangle other, double tolerance) {
```
can these be deleted?

ste7e replied:
Will do

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Utility/RANSACRegistration.cs:
```diff
@@ -0,0 +1,832 @@
+
+            public double[] AsPositionVector() {
+                return new double[] {
+                    normalizedPoints[0].X, normalizedPoints[0].Y, //normalizedPoints[0].NormalisedBrightness,
```
can the normalized brightnesses be deleted now?

ste7e replied:
Done

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Utility/RANSACRegistration.cs:
```diff
@@ -0,0 +1,832 @@
+            double BCy = c.Y - b.Y;
+
+            double dot = BAx * BCx + BAy * BCy;
+            double magBA = Math.Sqrt(BAx * BAx + BAy * BAy);
```
nit: can delay the square root until the division  to avoid an expensive operation for all stars

ste7e replied:
Got you.  Yes, moved the sqrt to work on the product now instead

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Utility/RANSACRegistration.cs:
```diff
@@ -0,0 +1,832 @@
+            var triangles = new List<StarTriangle>();
+            var pointsUsed = new HashSet<Point2D>();
+            int id = 0;
+            var brightnesses = point2Ds.Select(p => p.NormalisedBrightness).Order();
```
nit: it's wasteful to sort just to get min and max. that takes a 2N operation and turns it into an NLogN

ste7e replied:
Yep, agreed.  Fixed.

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Utility/RANSACRegistration.cs:
```diff
@@ -0,0 +1,832 @@
+                if (onePerPoint) {
+                    if (nearbyPoints.Count > 2) {
+                        double widestAngle = 0;
+                        int pt3 = 1;
```
what does pt3 represent? can you name this in a more descriptive way?

ste7e replied:
It's the third point of the triangle.  I've renamed it to thirdPoint

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Utility/RANSACRegistration.cs:
```diff
@@ -0,0 +1,832 @@
+                            var p3 = nearbyPoints[k];
+                            var triangle = new StarTriangle(imageSize, minBrightness, maxBrightness, pt, p2, p3, isReference, id++);
+                            triangles.Add(triangle);
+                            //pointsUsed.Add(pt);
```
remove

ste7e replied:
Done

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Utility/RANSACRegistration.cs:
```diff
@@ -0,0 +1,832 @@
+            return triangles;
+        }
+
+        private static double calcDistance(Point2D p1, Point2D p2) {
```
should this be in MathUtility?

ste7e replied:
Good idea.  Done

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Utility/RANSACRegistration.cs:
```diff
@@ -0,0 +1,832 @@
+            if (magnitudeA == 0 || magnitudeB == 0) {
+                return 0;
+            }
+            return dotProduct / (Math.Sqrt(magnitudeA) * Math.Sqrt(magnitudeB));
```
can do Math.Sqrt of their product instead

ste7e replied:
Done

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Utility/RANSACRegistration.cs:
```diff
@@ -0,0 +1,832 @@
+            return bestTransform;
+        }
+
+        public static Matrix3x2 EstimateAffineTransform(
```
i'm curious - is this an expensive operation?

ste7e replied:
Probably a little more than the EstimateSimilarityTransform.  It is needed though to cope with the different x and y scales

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Inspection/SensorModel.cs:
```diff
@@ -313,132 +332,307 @@ private List<SensorParaboloidDataPoint> ToInterpolatedGrid(
-                    foreach (var (star, starIndex) in result.StarDetectionResult.StarList.Select((star, starIndex) => (star, starIndex))) {
-                        tree.Add(new[] { star.Position.X, star.Position.Y }, new DetectedStarIndex(starIndex, (HocusFocusDetectedStar)star));
+                int maxStarsPerRegion = inspectorOptions.MaxStarsPerRegion;
+                // registration phase
```
can this be removed now?

ste7e replied:
Done

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/AutoFocus/InspectorVM.cs:
```diff
@@ -422,6 +408,149 @@ await Task.WhenAll(Enumerable.Range(0, FullSensorDetectedStars.Count).Select(i =
+                }
+
+                if ((trianglesByImage != null) && (trianglesByImage.Count > imageIndex)) {
+                    //foreach (var tri in trianglesByImage[i].Where(t => !t.IsReference && !t.Matched)) {
```
can this be removed now?

ste7e replied:
Done

ste7e committed:
Removal of commented out old code.
Removal of non-triangle-based RANSAC option
Removal of non-trees matching method
Removal of various commented out code chunkc

ste7e committed:
Added DISABLE_ILNUMERICS compile flag.  Add that flag to the conditional compilation symbols to dev on machine without ILN licence

ste7e committed:
Sort StarTriangle points in clockwise order

ste7e committed:
Made StarTriangle.Points publically read only

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/AutoFocus/AutoFocusEngine.cs:
```diff
@@ -1390,8 +1391,8 @@ private async Task<AutoFocusResult> RerunImpl(AutoFocusEngineOptions options, Sa
                     // Previous versions mistakenly saved the initial image in the attempt folder, which led to duplicate key exceptions
-                    var imageNumber = focuserPositionGroup.Max(g => g.ImageNumber);
-                    var files = focuserPositionGroup.Where(g => g.ImageNumber == imageNumber).OrderBy(g => g.FrameNumber).ToList();
+                    //var imageNumber = focuserPositionGroup.Max(g => g.ImageNumber);
```
can we remove this code now?

ste7e replied:
Absolutely.  Will take it out for the next commit

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Utility/NonLinearLeastSquaresSolver.cs:
```diff
@@ -86,7 +88,18 @@ public U SolveWinsorizedResiduals(
+
             while (disabledCount > 0 && winsorizedIterations++ < maxWinsorizedIterations) {
+                status.Progress++;
+                if (progress != null) {
```
nit: you can do progress?.Report(status) in one line

ste7e replied:
Yep, much neater.  Done

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Utility/RANSACRegistration.cs:
```diff
@@ -0,0 +1,783 @@
+                    var cosSim = CosineSimilarity(referenceTriangle.AsShapeVector(), imageTriangle.AsShapeVector());
+
+                    if (cosSim > bestCosSimLoc) {
+                        matches.Add((imageTriangle, cosSim));
```
hmm, this one isn't sitting well with me. if I am reading this right, you are doing a 2 pass scan and the first one building a list of potential matches based on shape cosine similarity but only as they get better. That means the order you scan in matters. what about taking the topN shape cosine similarity matches, and then picking the best based on brightness? or doing some combination of both and picking the best? like 50/50 from each?

ste7e replied:
Eugh!  That's an ugly mistake...  You're totally right.  That code at one point was built to collect all the matches better than the threshold.  I then changed it to just find find the best but obviously didn't pay enough attaention to what I was doing.  I've now changed it back to populate the matches collection for all matches better than the threshold.  Then if there are more than one match for any ref triangle, it goes to the brightness matching test.  As the threshold for a match based on location is quite high, it's not likely there will be many going into the brightness test phase so I don't think we need to limit the number to a topN set.

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Inspection/SensorModel.cs:
```diff
@@ -313,132 +332,287 @@ private List<SensorParaboloidDataPoint> ToInterpolatedGrid(
+                    ransacAligned = AlignStarsWithRANSAC(allDetectedStars, imageSize, stopwatch, ReferenceImage, progress);
+                    if (ransacAligned < allDetectedStars.Count) {
+                        Logger.Info("Ransac failed, reverting to non-aligned processing");
+                        //ResetRegistration(allDetectedStars);
```
Should this be uncommented? Also, this means a single bad frame would disable RANSAC. I think we should tolerate a bad frame or cases where there are below a minimum star count to harden this. Also if this fails, I think we should popup a notification so the user knows something suboptimal happened

ste7e replied:
No, I don't think we want to reset the registration of the aligned images.  It's something I toyed with and then decided against.  

So, what happens is, if all the frames get aligned by the RANSAC process, when we come to the search for matches across frames, we use a tighter search radius (as the images are aligned the stars should be prett6y close to the reference star).  On the other hand, if there are any frames which didn't get aligned, we throw that message into the log but all it really means is that when it comes to the looking for matches, we don't drop the search radius.  We still use the larger search radius because we have at least unaligned frame.

I'll delete the commented out line and try to make the logged message more meaningful.

ste7e replied:
I'd be happy to add some extra messaging to report back on the result of the alignment / star matching, though.

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Inspection/SensorModel.cs:
```diff
@@ -447,45 +621,522 @@ private RegistrationAndFitResult RegisterStarsAndFit(
+                    }
+                    int rejectedOnBrightness = globalNeighbors.Count() - queuedCount;
+                    //if (rejectedOnBrightness > 0)
+                    //    Logger.Info($"ImageStar: {imageIndex}/{sourceIndex}, {rejectedOnBrightness} matches out of {globalNeighbors.Count()} rejected on brightness difference (>{maxNormalisedBrightnessDiff})");
```
do you think this log would be helpful?

ste7e replied:
Yes, I've put it back in now

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Inspection/SensorModel.cs:
```diff
@@ -447,45 +621,522 @@ private RegistrationAndFitResult RegisterStarsAndFit(
+
+            int allStarCount = starDict.Count;
+            int minMatchPct = 10;   // proportion of images that must have star for star to be included
+            //double minBrightness = starDict.Keys.OrderBy(s => s.NormalisedBrightness).ElementAt((int)(allStarCount * .05)).NormalisedBrightness; // min set allow 95% of stars
```
should this code be removed?

ste7e replied:
Yep, tidied up

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Inspection/SensorModel.cs:
```diff
@@ -447,45 +621,522 @@ private RegistrationAndFitResult RegisterStarsAndFit(
+            return (registeredStars, totalRejectionsOnBrightness);
+        }
+
+        private (RegisteredStar[], int) MatchStarsWithoutKdTree(
```
can this whole method be removed now? are we using it anywhere?

ste7e replied:
It can and should be removed.  I was keeping it there in case we wanted to re-explore that option but I'm happy with the KdTree version now.  I'll clean it out

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Inspection/SensorModel.cs:
```diff
@@ -447,45 +621,522 @@ private RegistrationAndFitResult RegisterStarsAndFit(
+            return (registeredStars, rejectionsOnBrightness);
+        }
+
+        private static int MatchStarsByNN(
```
same with this one?

ste7e replied:
Agreed.  Done

ghilios commented on Joko.NINA.Plugins/Joko.NINA.Plugins.HocusFocus/Inspection/SensorModel.cs:
```diff
@@ -447,45 +621,522 @@ private RegistrationAndFitResult RegisterStarsAndFit(
+            return ret;
+        }
+
+        private static double DistanceSquared(float[] p1, float[] p2) {
```
can we use the MathUtility version instead?

ste7e replied:
Yep, this method isn't being used now having got rid the non-kdTree code.  Removed now.

ste7e committed:
Removed some unused code.

Fixed bug where putative match generation was adding too many 'matches' to matches collection.

ste7e committed:
First go at adding a registrationAndFitReport to pass warnings about failed RANSAC etc back to the user

ste7e committed:
Added a bit more messaging

ste7e committed:
Default useRANSAC to false
Save alignment images on live run as well as re-runs
Save the alignment transform matrix and print this out to the alignment and registration images
